File: app/src/main/java/com/expansion/lg/kimaru/training/network/TrainingDataSync.java
package com.expansion.lg.kimaru.training.network;

import android.content.Context;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.expansion.lg.kimaru.training.database.DatabaseHelper;
import com.expansion.lg.kimaru.training.objs.Training;
import com.expansion.lg.kimaru.training.objs.TrainingExam;
import com.expansion.lg.kimaru.training.objs.TrainingExamResult;
import com.expansion.lg.kimaru.training.utils.Constants;
import com.expansion.lg.kimaru.training.utils.SessionManagement;
import com.kimarudg.async.http.AsyncHttpClient;
import com.kimarudg.async.http.AsyncHttpPost;
import com.kimarudg.async.http.body.JSONArrayBody;
import com.kimarudg.async.http.body.JSONObjectBody;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

/**
 * Created by kimaru on 2/27/18.
 */

public class TrainingDataSync {
    Context context;
    DatabaseHelper databaseHelper;
    SessionManagement session;

    public TrainingDataSync(Context context){
        this.context = context;
        this.databaseHelper = new DatabaseHelper(context);
        session = new SessionManagement(context);
    }

    public void pollNewTrainings(){
        new syncTrainings().execute();
    }

    public void startPollNewTrainingsTask(){

        final Handler handler = new Handler(Looper.getMainLooper());
        Timer timer = new Timer();
        TimerTask getTrainingsTask = new TimerTask() {
            @Override
            public void run() {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        new syncTrainings().execute();
                    }
                });
            }
        };
        timer.schedule(getTrainingsTask, 0, 60*500 * 1); //every 30 minutes
    }

    public void startPollUploadExamResults(final String trainingId){

        final Handler handler = new Handler(Looper.getMainLooper());
        Timer timer = new Timer();
        TimerTask getTrainingsTask = new TimerTask() {
            @Override
            public void run() {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        new UploadCertificationsResults().execute(trainingId);
                    }
                });
            }
        };
        timer.schedule(getTrainingsTask, 0, 60*500 * 1);
    }

    private class syncTrainings extends AsyncTask<Void, Void, Void> {
        @Override
        protected void onPreExecute(){
            super.onPreExecute();
        }

        @Override
        protected Void doInBackground(Void... arg0){
            JsonParser jsonParser = new JsonParser();
            SessionManagement sessionManagement = new SessionManagement(context);

            String json = jsonParser.getJSONFromUrl(sessionManagement.getApiUrl()
                    +sessionManagement.getTrainingEndpoint());
            if(json !=null){
                try{
                    JSONObject reader= new JSONObject(json);
                    JSONArray recs = reader.getJSONArray(sessionManagement.getTrainingJSONRoot());
                    for (int x = 0; x < recs.length(); x++){
                        Log.d("Tremap", "===================");
                        Log.d("Tremap", recs.getString(x));
                        Log.d("Tremap", "===================");
                        databaseHelper.trainingFromJson(recs.getJSONObject(x));
                    }
                }catch(JSONException e){
                    Log.d("TREMAP", "TRAINING ERROR "+ e.getMessage());
                }
            }
            return null;
        }
        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);
        }
    }
    /**
     * TRAINEES
     *
     */
    public void pollNewTrainingTrainees(){
        new syncTrainingTrainees().execute();
    }
    public void startPollNewTraineesTask(){

        final Handler handler = new Handler(Looper.getMainLooper());
        Timer timer = new Timer();
        TimerTask getTrainingTraineesTask = new TimerTask() {
            @Override
            public void run() {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        new syncTrainingTrainees().execute();
                    }
                });
            }
        };
        timer.schedule(getTrainingTraineesTask, 0, 60*500 * 1); //every 30 minutes
    }
    private class syncTrainingTrainees extends AsyncTask<Void, Void, Void> {
        @Override
        protected void onPreExecute(){
            super.onPreExecute();
        }

        @Override
        protected Void doInBackground(Void... arg0){
            JsonParser jsonParser = new JsonParser();
            SessionManagement sessionManagement = new SessionManagement(context);

            String json = jsonParser.getJSONFromUrl(sessionManagement.getApiUrl()
                    +sessionManagement.getTrainingTraineesEndpoint());
            if(json !=null){
                try{
                    JSONObject reader= new JSONObject(json);
                    JSONArray recs = reader.getJSONArray(sessionManagement.getTrainingJSONRoot());
                    for (int x = 0; x < recs.length(); x++){
                        databaseHelper.trainingTraineeFromJson(recs.getJSONObject(x));
                    }
                }catch(JSONException e){
                    Log.d("TREMAP", "TRAINEE ERROR "+ e.getMessage());
                }
            }
            return null;
        }
        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);
        }
    }


    public void getTrainingDetailsJson(String trainingId){
        //new syncTrainings().execute();
        new getTrainingDetails().execute(trainingId);
    }

    private class getTrainingDetails extends AsyncTask<String, Void, String>{
        @Override
        protected void onPreExecute(){
            super.onPreExecute();
        }
        @Override
        protected String doInBackground(String... strings){
            String trainigId = strings[0];
            JsonParser jsonParser = new JsonParser();
            SessionManagement sessionManagement = new SessionManagement(context);
            String url = sessionManagement
                    .getApiUrl()+sessionManagement.getTrainingDetailsEndpoint()+"/"+trainigId;

            String json = jsonParser.getJSONFromUrl(url);
            if (json != null){
                try{
                    JSONObject training = new JSONObject(json).getJSONObject(sessionManagement.getTrainingDetailsJsonRoot());
                    int x;
                    if (!training.isNull("classes")) {
                        JSONArray trainingClassesJson = training.getJSONArray("classes");
                        for (x = 0; x < trainingClassesJson.length(); x++) {
                            databaseHelper.trainingClassFromJson(trainingClassesJson.getJSONObject(x));
                        }
                    }

                    if (!training.isNull("trainees")) {
                        JSONArray trainingTrainees = training.getJSONArray("trainees");
                        for (x=0; x<trainingTrainees.length(); x++){
                            databaseHelper.trainingTraineeFromJson(trainingTrainees.getJSONObject(x));
                        }
                    }

                    if (!training.isNull("training_sessions")) {
                        JSONArray trainingSessions = training.getJSONArray("training_sessions");
                        for (x = 0; x < trainingSessions.length(); x++) {
                            databaseHelper.trainingSessionFromJson(trainingSessions.getJSONObject(x));

                        }
                    }
                    if (!training.isNull("training_venue_details")) {
                        JSONObject trainingVenueDetails = training.getJSONObject("training_venue_details");
                        databaseHelper.trainingVenueFromJson(trainingVenueDetails);
                    }


                }catch (JSONException e){
                    Log.d("Tremap", "-----------ERROR----------------");
                    Log.d("Tremap", e.getMessage());
                }

            }else {
                Log.d("Tremap", "--------NULL JSON-------------------");
            }
            return null;
        }
        @Override
        protected void onPostExecute(String result) {
        }
    }


    //Session Topics
    public void syncSessionTopics(){
        new getTrainingTopics().execute();
    }
    private class getTrainingTopics extends AsyncTask<Void, Void, Void>{
        @Override
        protected void onPreExecute(){
            super.onPreExecute();
        }
        @Override
        protected Void doInBackground(Void... args0){
            JsonParser jsonParser = new JsonParser();
            SessionManagement sessionManagement = new SessionManagement(context);
            String url = sessionManagement
                    .getApiUrl()+sessionManagement.getSessionTopicsEndpoint();
            String json = jsonParser.getJSONFromUrl(url);
            Log.d("Tremap", json);
            if (json != null){
                try{
                    JSONArray sessionTopics = new JSONObject(json).getJSONArray(sessionManagement.getSessionTopicsJsonRoot());
                    for (int x = 0; x < sessionTopics.length(); x++){
                        databaseHelper.sessionTopicFromJson(sessionTopics.getJSONObject(x));
                    }
                }catch (JSONException e){
                    Log.d("Tremap", "-----------ERROR----------------");
                    Log.d("Tremap", e.getMessage());
                }

            }else {
                Log.d("Tremap", "--------NULL JSON-------------------");
            }
            return null;
        }
        @Override
        protected void onPostExecute(Void result) {
        }
    }



    public void getSessionAttendance(String trainingId){
        new getSessionAttendances().execute(trainingId);
    }

    private class getSessionAttendances extends AsyncTask<String, Void, String>{
        @Override
        protected void onPreExecute(){
            super.onPreExecute();
        }

        @Override
        protected String doInBackground(String... strings){
            String trainingId = strings[0];
            JsonParser jsonParser = new JsonParser();
            SessionManagement sessionManagement = new SessionManagement(context);
            String url = sessionManagement.getApiUrl()+sessionManagement
                    .getSessionAttendanceEndpoint() + "/"+trainingId;
            String json = jsonParser.getJSONFromUrl(url);
            if (json != null){
                try{
                    JSONArray attendances = new JSONObject(json).getJSONArray(sessionManagement.getSessionAttendanceJSONRoot());
                    for (int x = 0; x <attendances.length(); x++){
                        databaseHelper.sessionAttendanceFromJSON(attendances.getJSONObject(x));
                    }
                }catch (Exception e){
                    Log.d("Tremap","///////////////////////////////");
                    Log.d("Tremap", e.getMessage());
                    Log.d("Tremap","///////////////////////////////");
                }
            }
            return null;
        }

        @Override
        protected void onPostExecute(String result){}
    }


    public void getTraineeStatusFromCloud(){
        new getTraineeStatusFromUpstream().execute();
    }

    private class getTraineeStatusFromUpstream extends AsyncTask<Void, Void, Void>{
        @Override
        protected void onPreExecute(){super.onPreExecute();}

        @Override
        protected Void doInBackground(Void... voids){
            JsonParser jsonParser = new JsonParser();
            SessionManagement session =  new SessionManagement(context);
            String url = session.getApiUrl()+session.getTraineeStatusEndpoint();
            Log.d("Tremap", "____________________---URL-------------____________");
            Log.d("Tremap", url);
            Log.d("Tremap", "____________________----------------____________");

            String json = jsonParser.getJSONFromUrl(url);
            if (json != null){
                try{
                    JSONArray statuses = new JSONObject(json).getJSONArray(session.getTraineeStatusJSONRoot());
                    for (int x=0; x < statuses.length(); x++){
                        databaseHelper.traineeStatusFromJson(statuses.getJSONObject(x));
                    }
                }catch (Exception e){
                    Log.d("Tremap", "____________________----------------____________");
                    Log.d("Tremap", e.getMessage());
                    Log.d("Tremap", "____________________----------------____________");
                }
            }
            return null;
        }
    }


    public void getTrainingExams(String trainingId){
        new getTrainingExamsFromUpstream().execute(trainingId);
        new GetTrainingCertifications().execute(trainingId);
    }

    private class GetTrainingCertifications extends AsyncTask<String, Void, String> {
        protected String doInBackground(String ...strings) {
            JsonParser jsonParser = new JsonParser();
            String trainingId = strings[0];
            SessionManagement session =  new SessionManagement(context);
            String url = session.getApiUrl()+ String.format(session.getTrainingCertificationsEndpoint(), trainingId);
            Log.d("TREMAP", url);
            Log.d("TREMAP", url);
            Log.d("TREMAP", url);

            String json = jsonParser.getJSONFromUrl(url);
            if (json != null){
                try{
                    JSONArray exams = new JSONObject(json).getJSONArray(session.getTrainingExamsJSONRoot());
                    for (int x=0; x < exams.length(); x++){
                        databaseHelper.trainingExamFromJson(exams.getJSONObject(x));
                    }
                }catch (Exception e){
                    Log.d("TREMAPEXAMS", "ERROR TRAINING EXAMS");
                    Log.d("Tremap", e.getMessage());
                }
            }
            return null;
        }
    }

    private class getTrainingExamsFromUpstream extends AsyncTask<String, Void, String>{
        @Override
        protected void onPreExecute(){super.onPreExecute();}

        @Override
        protected String doInBackground(String... strings){
            JsonParser jsonParser = new JsonParser();
            String trainingId = strings[0];
            SessionManagement session =  new SessionManagement(context);
            String url = session.getApiUrl()+ String.format(session.getTrainingExamsEndpoint(), trainingId);
            Log.d("TREMAP", url);
            Log.d("TREMAP", url);
            Log.d("TREMAP", url);

            String json = jsonParser.getJSONFromUrl(url);
            if (json != null){
                try{
                    JSONArray exams = new JSONObject(json).getJSONArray(session.getTrainingExamsJSONRoot());
                    for (int x=0; x < exams.length(); x++){
                        databaseHelper.trainingExamFromJson(exams.getJSONObject(x));
                    }
                }catch (Exception e){
                    Log.d("TREMAPEXAMS", "ERROR TRAINING EXAMS");
                    Log.d("Tremap", e.getMessage());
                }
            }
            return null;
        }
    }


    public void getTrainingExamResults(String examId){
        new getTrainingExamResultsFromUpstream().execute(examId);
    }

    private class getTrainingExamResultsFromUpstream extends AsyncTask<String, Void, String>{
        @Override
        protected void onPreExecute(){super.onPreExecute();}

        @Override
        protected String doInBackground(String... strings){
            JsonParser jsonParser = new JsonParser();
            String examId = strings[0];
            SessionManagement session =  new SessionManagement(context);
            String url = session.getApiUrl()+ String.format(session.getTrainingExamResultsEndpoint(), examId);
            Log.d("Tremap", "************************************************************************");
            Log.d("Tremap", url);
            Log.d("Tremap", "************************************************************************");
            String json = jsonParser.getJSONFromUrl(url);
            if (json != null){
                try{
                    JSONArray results = new JSONObject(json).getJSONArray(session.getTrainingExamResultsJSONRoot());
                    for (int x=0; x < results.length(); x++){
                        databaseHelper.trainingExamResultFromJson(results.getJSONObject(x));
                    }
                }catch (Exception e){
                    Log.d("TREMAPEXAMS", "ERROR TRAINING EXAM RESULTS");
                    Log.d("Tremap", e.getMessage());
                }
            }
            return null;
        }
    }

    public void startUploadAttendanceTask(){
        final Handler handler = new Handler(Looper.getMainLooper());
        Timer timer = new Timer();
        TimerTask getTrainingsTask = new TimerTask() {
            @Override
            public void run() {
                handler.post(new Runnable() {
                    @Override
                    public void run()     {
                        new uploadSessionAttendanceToUpstream().execute("all");
                    }
                });
            }
        };
        timer.schedule(getTrainingsTask, 0, 600000 * 60); // every one hour
    }
    //SessionAttendance
    public void uploadSessionAttendance(String trainingId){
        JSONObject json =databaseHelper.trainingSessionAttendanceToUploadJson(trainingId);
        try{
            String syncResults = syncClient(json,
                    session.getUploadSessionAttendanceEndpoint()+"/"+trainingId);
        }catch (Exception e){}
    }

    private class UploadCertificationsResults extends AsyncTask<String, Void, String> {

        @Override
        protected String doInBackground(String... strings) {

            String trainingId = strings[0];
            DatabaseHelper db = new DatabaseHelper(context);
            String endpoint = session.getApiUrl() + "training/exam/result/save";

            List<TrainingExam> exams = db.getTrainingExamsByTrainingId(trainingId);

            try {

                JSONArray params = new JSONArray();
                for (TrainingExam exam: exams) {
                    if (exam.getCertificationTypeId() > 0) {
                        List<TrainingExamResult> results = db.getTrainingExamResultsByExam(exam.getId().toString());
                        for (TrainingExamResult result: results) {
                            JSONObject param = new JSONObject();


                            param.put("training_exam_id", result.getTrainingExamId());
                            param.put("trainee_id", result.getTraineeId());
                            param.put("question_id", result.getQuestionId());
                            param.put("question_score", result.getQuestionScore());
                            param.put("country", result.getCountry());
                            param.put("answer", result.getAnswer());
                            param.put("choice_id", result.getChoiceId());
                            param.put("id", result.getId());

                            params.put(param);
                        }

                    }
                }

                AsyncHttpPost p = new AsyncHttpPost(endpoint);
                Log.e("Tremap", params.toString());
                p.setBody(new JSONArrayBody(params));
                JSONObject ret = AsyncHttpClient.getDefaultInstance().executeJSONObject(p, null).get();
                Log.e("RESULTS : Sync", ret.toString());

            } catch (Exception ex) {

                ex.printStackTrace();
            }

            return null;
        }
    }

    private class uploadSessionAttendanceToUpstream extends AsyncTask<String, Void, String>{
        @Override
        protected void onPreExecute(){super.onPreExecute();}

        @Override
        protected String doInBackground(String... strings){
            String trainingId = strings[0];
            String syncResults;
            JSONObject json;
            try{
                if (strings[0].equalsIgnoreCase("all")){
                    json = databaseHelper.allTrainingSessionAttendanceToJson();
                    syncResults = syncClient(json, session.getUploadSessionAttendanceEndpoint());
                }else{
                    json =databaseHelper.trainingSessionAttendanceToUploadJson(trainingId);
                    syncResults = syncClient(json,
                            session.getUploadSessionAttendanceEndpoint()+"/"+trainingId);
                }
            }catch (Exception e){
                syncResults = null;
            }
            return syncResults;
        }
    }

    // Callback for the API
    private String syncClient(JSONObject json, String apiEndpoint) throws Exception {
        SessionManagement session = new SessionManagement(context);
        String url = session.getApiUrl()+apiEndpoint;
        Log.d("RESULTS : Sync", "+++++++++++++++++++++++++++");
        Log.d("RESULTS : Sync", url);
        Log.d("RESULTS : Sync", "+++++++++++++++++++++++++++");
        AsyncHttpPost p = new AsyncHttpPost(url);
        p.setBody(new JSONObjectBody(json));
        JSONObject ret = AsyncHttpClient.getDefaultInstance().executeJSONObject(p, null).get();
//        return ret.getString(expectedJsonRoot);
        Log.d("RESULTS : Sync", ret.toString());
        return ret.toString();
    }
}

File: app/src/main/java/com/expansion/lg/kimaru/training/network/TrainingDataSync.kt
package com.expansion.lg.kimaru.training.network

import android.content.Context
import android.os.AsyncTask
import android.os.Handler
import android.os.Looper
import android.util.Log

import com.expansion.lg.kimaru.training.database.DatabaseHelper
import com.expansion.lg.kimaru.training.objs.Training
import com.expansion.lg.kimaru.training.utils.Constants
import com.expansion.lg.kimaru.training.utils.SessionManagement
import com.kimarudg.async.http.AsyncHttpClient
import com.kimarudg.async.http.AsyncHttpPost
import com.kimarudg.async.http.body.JSONObjectBody

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList
import java.util.Timer
import java.util.TimerTask

/**
 * Created by kimaru on 2/27/18.
 */

class TrainingDataSync(internal var context: Context) {
    internal var databaseHelper: DatabaseHelper
    internal var session: SessionManagement

    init {
        this.databaseHelper = DatabaseHelper(context)
        session = SessionManagement(context)
    }

    fun pollNewTrainings() {
        syncTrainings().execute()
    }


    fun startPollNewTrainingsTask() {

        val handler = Handler(Looper.getMainLooper())
        val timer = Timer()
        val getTrainingsTask = object : TimerTask() {
            override fun run() {
                handler.post { syncTrainings().execute() }
            }
        }
        timer.schedule(getTrainingsTask, 0, (60 * 500 * 1).toLong()) //every 30 minutes
    }


    private inner class syncTrainings : AsyncTask<Void, Void, Void>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg arg0: Void): Void? {
            val jsonParser = JsonParser()
            val sessionManagement = SessionManagement(context)

            val json = jsonParser.getJSONFromUrl(sessionManagement.apiUrl + sessionManagement.trainingEndpoint)
            if (json != null) {
                try {
                    val reader = JSONObject(json)
                    val recs = reader.getJSONArray(sessionManagement.trainingJSONRoot)
                    for (x in 0 until recs.length()) {
                        Log.d("Tremap", "===================")
                        Log.d("Tremap", recs.getString(x))
                        Log.d("Tremap", "===================")
                        databaseHelper.trainingFromJson(recs.getJSONObject(x))
                    }
                } catch (e: JSONException) {
                    Log.d("TREMAP", "TRAINING ERROR " + e.message)
                }

            }
            return null
        }

        override fun onPostExecute(result: Void) {
            super.onPostExecute(result)
        }
    }

    /**
     * TRAINEES
     *
     */
    fun pollNewTrainingTrainees() {
        syncTrainingTrainees().execute()
    }

    fun startPollNewTraineesTask() {

        val handler = Handler(Looper.getMainLooper())
        val timer = Timer()
        val getTrainingTraineesTask = object : TimerTask() {
            override fun run() {
                handler.post { syncTrainingTrainees().execute() }
            }
        }
        timer.schedule(getTrainingTraineesTask, 0, (60 * 500 * 1).toLong()) //every 30 minutes
    }

    private inner class syncTrainingTrainees : AsyncTask<Void, Void, Void>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg arg0: Void): Void? {
            val jsonParser = JsonParser()
            val sessionManagement = SessionManagement(context)

            val json = jsonParser.getJSONFromUrl(sessionManagement.apiUrl + sessionManagement.trainingTraineesEndpoint)
            if (json != null) {
                try {
                    val reader = JSONObject(json)
                    val recs = reader.getJSONArray(sessionManagement.trainingJSONRoot)
                    for (x in 0 until recs.length()) {
                        databaseHelper.trainingTraineeFromJson(recs.getJSONObject(x))
                    }
                } catch (e: JSONException) {
                    Log.d("TREMAP", "TRAINEE ERROR " + e.message)
                }

            }
            return null
        }

        override fun onPostExecute(result: Void) {
            super.onPostExecute(result)
        }
    }


    fun getTrainingDetailsJson(trainingId: String) {
        //new syncTrainings().execute();
        getTrainingDetails().execute(trainingId)
    }

    private inner class getTrainingDetails : AsyncTask<String, Void, String>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg strings: String): String? {
            val trainigId = strings[0]
            val jsonParser = JsonParser()
            val sessionManagement = SessionManagement(context)
            val url = sessionManagement
                    .apiUrl + sessionManagement.trainingDetailsEndpoint + "/" + trainigId

            val json = jsonParser.getJSONFromUrl(url)
            if (json != null) {
                try {
                    val training = JSONObject(json).getJSONObject(sessionManagement.trainingDetailsJsonRoot)
                    var x: Int
                    if (!training.isNull("classes")) {
                        val trainingClassesJson = training.getJSONArray("classes")
                        x = 0
                        while (x < trainingClassesJson.length()) {
                            databaseHelper.trainingClassFromJson(trainingClassesJson.getJSONObject(x))
                            x++
                        }
                    }

                    if (!training.isNull("trainees")) {
                        val trainingTrainees = training.getJSONArray("trainees")
                        x = 0
                        while (x < trainingTrainees.length()) {
                            databaseHelper.trainingTraineeFromJson(trainingTrainees.getJSONObject(x))
                            x++
                        }
                    }

                    if (!training.isNull("training_sessions")) {
                        val trainingSessions = training.getJSONArray("training_sessions")
                        x = 0
                        while (x < trainingSessions.length()) {
                            databaseHelper.trainingSessionFromJson(trainingSessions.getJSONObject(x))
                            x++

                        }
                    }
                    if (!training.isNull("training_venue_details")) {
                        val trainingVenueDetails = training.getJSONObject("training_venue_details")
                        databaseHelper.trainingVenueFromJson(trainingVenueDetails)
                    }


                } catch (e: JSONException) {
                    Log.d("Tremap", "-----------ERROR----------------")
                    Log.d("Tremap", e.message)
                }

            } else {
                Log.d("Tremap", "--------NULL JSON-------------------")
            }
            return null
        }

        override fun onPostExecute(result: String) {}
    }


    //Session Topics
    fun syncSessionTopics() {
        getTrainingTopics().execute()
    }

    private inner class getTrainingTopics : AsyncTask<Void, Void, Void>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg args0: Void): Void? {
            val jsonParser = JsonParser()
            val sessionManagement = SessionManagement(context)
            val url = sessionManagement
                    .apiUrl + sessionManagement.sessionTopicsEndpoint
            val json = jsonParser.getJSONFromUrl(url)
            Log.d("Tremap", json)
            if (json != null) {
                try {
                    val sessionTopics = JSONObject(json).getJSONArray(sessionManagement.sessionTopicsJsonRoot)
                    for (x in 0 until sessionTopics.length()) {
                        databaseHelper.sessionTopicFromJson(sessionTopics.getJSONObject(x))
                    }
                } catch (e: JSONException) {
                    Log.d("Tremap", "-----------ERROR----------------")
                    Log.d("Tremap", e.message)
                }

            } else {
                Log.d("Tremap", "--------NULL JSON-------------------")
            }
            return null
        }

        override fun onPostExecute(result: Void) {}
    }


    fun getSessionAttendance(trainingId: String) {
        getSessionAttendances().execute(trainingId)
    }

    private inner class getSessionAttendances : AsyncTask<String, Void, String>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg strings: String): String? {
            val trainingId = strings[0]
            val jsonParser = JsonParser()
            val sessionManagement = SessionManagement(context)
            val url = sessionManagement.apiUrl + sessionManagement
                    .sessionAttendanceEndpoint + "/" + trainingId
            val json = jsonParser.getJSONFromUrl(url)
            if (json != null) {
                try {
                    val attendances = JSONObject(json).getJSONArray(sessionManagement.sessionAttendanceJSONRoot)
                    for (x in 0 until attendances.length()) {
                        databaseHelper.sessionAttendanceFromJSON(attendances.getJSONObject(x))
                    }
                } catch (e: Exception) {
                    Log.d("Tremap", "///////////////////////////////")
                    Log.d("Tremap", e.message)
                    Log.d("Tremap", "///////////////////////////////")
                }

            }
            return null
        }

        override fun onPostExecute(result: String) {}
    }


    fun getTraineeStatusFromCloud() {
        getTraineeStatusFromUpstream().execute()
    }

    private inner class getTraineeStatusFromUpstream : AsyncTask<Void, Void, Void>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg voids: Void): Void? {
            val jsonParser = JsonParser()
            val session = SessionManagement(context)
            val url = session.apiUrl + session.traineeStatusEndpoint
            Log.d("Tremap", "____________________---URL-------------____________")
            Log.d("Tremap", url)
            Log.d("Tremap", "____________________----------------____________")

            val json = jsonParser.getJSONFromUrl(url)
            if (json != null) {
                try {
                    val statuses = JSONObject(json).getJSONArray(session.traineeStatusJSONRoot)
                    for (x in 0 until statuses.length()) {
                        databaseHelper.traineeStatusFromJson(statuses.getJSONObject(x))
                    }
                } catch (e: Exception) {
                    Log.d("Tremap", "____________________----------------____________")
                    Log.d("Tremap", e.message)
                    Log.d("Tremap", "____________________----------------____________")
                }

            }
            return null
        }
    }


    fun getTrainingExams(trainingId: String) {
        getTrainingExamsFromUpstream().execute(trainingId)
    }

    private inner class getTrainingExamsFromUpstream : AsyncTask<String, Void, String>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg strings: String): String? {
            val jsonParser = JsonParser()
            val trainingId = strings[0]
            val session = SessionManagement(context)
            val url = session.apiUrl + String.format(session.trainingExamsEndpoint, trainingId)
            Log.d("TREMAP", url)
            Log.d("TREMAP", url)
            Log.d("TREMAP", url)

            val json = jsonParser.getJSONFromUrl(url)
            if (json != null) {
                try {
                    val exams = JSONObject(json).getJSONArray(session.trainingExamsJSONRoot)
                    for (x in 0 until exams.length()) {
                        databaseHelper.trainingExamFromJson(exams.getJSONObject(x))
                    }
                } catch (e: Exception) {
                    Log.d("TREMAPEXAMS", "ERROR TRAINING EXAMS")
                    Log.d("Tremap", e.message)
                }

            }
            return null
        }
    }


    fun getTrainingExamResults(examId: String) {
        getTrainingExamResultsFromUpstream().execute(examId)
    }

    private inner class getTrainingExamResultsFromUpstream : AsyncTask<String, Void, String>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg strings: String): String? {
            val jsonParser = JsonParser()
            val examId = strings[0]
            val session = SessionManagement(context)
            val url = session.apiUrl + String.format(session.trainingExamResultsEndpoint, examId)
            Log.d("Tremap", "************************************************************************")
            Log.d("Tremap", url)
            Log.d("Tremap", "************************************************************************")
            val json = jsonParser.getJSONFromUrl(url)
            if (json != null) {
                try {
                    val results = JSONObject(json).getJSONArray(session.trainingExamResultsJSONRoot)
                    for (x in 0 until results.length()) {
                        databaseHelper.trainingExamResultFromJson(results.getJSONObject(x))
                    }
                } catch (e: Exception) {
                    Log.d("TREMAPEXAMS", "ERROR TRAINING EXAM RESULTS")
                    Log.d("Tremap", e.message)
                }

            }
            return null
        }
    }

    fun startUploadAttendanceTask() {
        val handler = Handler(Looper.getMainLooper())
        val timer = Timer()
        val getTrainingsTask = object : TimerTask() {
            override fun run() {
                handler.post { uploadSessionAttendanceToUpstream().execute("all") }
            }
        }
        timer.schedule(getTrainingsTask, 0, (600000 * 60).toLong()) // every one hour
    }

    //SessionAttendance
    fun uploadSessionAttendance(trainingId: String) {
        val json = databaseHelper.trainingSessionAttendanceToUploadJson(trainingId)
        try {
            val syncResults = syncClient(json,
                    session.uploadSessionAttendanceEndpoint + "/" + trainingId)
        } catch (e: Exception) {
        }

    }

    private inner class uploadSessionAttendanceToUpstream : AsyncTask<String, Void, String>() {
        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg strings: String): String? {
            val trainingId = strings[0]
            var syncResults: String?
            val json: JSONObject
            try {
                if (strings[0].equals("all", ignoreCase = true)) {
                    json = databaseHelper.allTrainingSessionAttendanceToJson()
                    syncResults = syncClient(json, session.uploadSessionAttendanceEndpoint)
                } else {
                    json = databaseHelper.trainingSessionAttendanceToUploadJson(trainingId)
                    syncResults = syncClient(json,
                            session.uploadSessionAttendanceEndpoint + "/" + trainingId)
                }
            } catch (e: Exception) {
                syncResults = null
            }

            return syncResults
        }
    }

    // Callback for the API
    @Throws(Exception::class)
    private fun syncClient(json: JSONObject, apiEndpoint: String): String {
        val session = SessionManagement(context)
        val url = session.apiUrl + apiEndpoint
        Log.d("RESULTS : Sync", "+++++++++++++++++++++++++++")
        Log.d("RESULTS : Sync", url)
        Log.d("RESULTS : Sync", "+++++++++++++++++++++++++++")
        val p = AsyncHttpPost(url)
        p.body = JSONObjectBody(json)
        val ret = AsyncHttpClient.getDefaultInstance().executeJSONObject(p, null).get()
        //        return ret.getString(expectedJsonRoot);
        Log.d("RESULTS : Sync", ret.toString())
        return ret.toString()
    }
}